pkg/controller/kfservice/resources/istio/virtualservice.go

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2019 kubeflow.org.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package istio
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha2"
+	"github.com/kubeflow/kfserving/pkg/constants"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+	istiov1alpha1 "knative.dev/pkg/apis/istio/common/v1alpha1"
+	istiov1alpha3 "knative.dev/pkg/apis/istio/v1alpha3"
+)
+
+type VirtualServiceBuilder struct {
+}
+
+func NewVirtualServiceBuilder() *VirtualServiceBuilder {
+	return &VirtualServiceBuilder{}
+}
+
+func createFailedStatus(reason string, message string) *v1alpha2.VirtualServiceStatus {
+	return &v1alpha2.VirtualServiceStatus{
+		Status: duckv1beta1.Status{
+			Conditions: duckv1beta1.Conditions{{
+				Type:    v1alpha2.RoutesReady,
+				Status:  corev1.ConditionFalse,
+				Reason:  reason,
+				Message: message,
+			}},
+		},
+	}
+}
+
+func (r *VirtualServiceBuilder) CreateVirtualService(kfsvc *v1alpha2.KFService) (*istiov1alpha3.VirtualService, *v1alpha2.VirtualServiceStatus) {
+
```

Comment:
@yuzisun can take a look at this and give feedback?

Reply:
See example vs here: https://github.com/rakelkar/examples/blob/master/istio/kn-kf-example

Reply:
in `hosts` section, we may also need to add internal domains like `mykfsvc.default.svc.cluster.local` in case it is called from in cluster.

Reply:
Opened a separate issue for this (this PR has gone on too long!)
https://github.com/kubeflow/kfserving/issues/412